tests/handler/handler_copy.cpp

```diff
@@ -17,334 +22,887 @@ using mode_t = cl::sycl::access::mode;
-                                 target_t::global_buffer>(queue);
-    test_write_acc_copy_functions<dataT, dim_src, dim_dst, mode_t::read_write,
-                                  mode_t::write, target_t::global_buffer>(
-        queue);
-    test_write_acc_copy_functions<dataT, dim_src, dim_dst, mode_t::read_write,
-                                  mode_t::read_write, target_t::global_buffer>(
-        queue);
-    test_write_acc_copy_functions<dataT, dim_src, dim_dst, mode_t::read_write,
-                                  mode_t::discard_write,
-                                  target_t::global_buffer>(queue);
-    test_write_acc_copy_functions<dataT, dim_src, dim_dst, mode_t::read_write,
-                                  mode_t::discard_read_write,
-                                  target_t::global_buffer>(queue);
+    constexpr auto target = target_t::global_buffer;
+    test_read_acc_copy_functions<dataT, dim_src, mode_t::read, target, strided,
+                                 transposed>(lh, queue);
+    test_read_acc_copy_functions<dataT, dim_src, mode_t::read_write, target,
+                                 strided, transposed>(lh, queue);
   }
   {
-    // target == constant_buffer
-    test_read_acc_copy_functions<dataT, dim_src, mode_t::read,
-                                 target_t::constant_buffer>(queue);
+    constexpr auto target = target_t::constant_buffer;
+    test_read_acc_copy_functions<dataT, dim_src, mode_t::read, target, strided,
+                                 transposed>(lh, queue);
   }
 }
 
-template <int dim_src, int dim_dst>
-void test_all_dimensions(cl::sycl::queue& queue) {
-  test_all_copy_functions<char, dim_src, dim_dst>(queue);
-  test_all_copy_functions<short, dim_src, dim_dst>(queue);
-  test_all_copy_functions<int, dim_src, dim_dst>(queue);
-  test_all_copy_functions<long, dim_src, dim_dst>(queue);
-  test_all_copy_functions<float, dim_src, dim_dst>(queue);
-  test_all_copy_functions<double, dim_src, dim_dst>(queue);
-
-  test_all_copy_functions<cl::sycl::char2, dim_src, dim_dst>(queue);
-  test_all_copy_functions<cl::sycl::short3, dim_src, dim_dst>(queue);
-  test_all_copy_functions<cl::sycl::int4, dim_src, dim_dst>(queue);
-  test_all_copy_functions<cl::sycl::long8, dim_src, dim_dst>(queue);
-  test_all_copy_functions<cl::sycl::float8, dim_src, dim_dst>(queue);
-  test_all_copy_functions<cl::sycl::double16, dim_src, dim_dst>(queue);
+/**
+ * @brief Tests all valid combinations of source and destination access modes.
+ */
+template <typename dataT, int dim_src, int dim_dst, bool strided,
+          bool transposed>
+void test_all_write_acc_copy_functions(log_helper lh, cl::sycl::queue& queue) {
+  lh = lh.set_dim_src(dim_src).set_dim_dst(dim_dst);
+  constexpr auto target = target_t::global_buffer;
+
+  constexpr auto st = strided;
+  constexpr auto tr = transposed;
+
+  test_write_acc_copy_functions<dataT, dim_src, dim_dst, mode_t::read,
```

Comment:
```c++
for(auto s_m : { mode_t::read, mode_t::read_write })
  for (auto d_m : { mode_t::write, mode_t::read_write, mode_t::discard_write, mode_t::discard_read_write })
...
```

Reply:
That would be nice, yes, but I'd need a `for constexpr` for that :-). Once the CTS upgrades to C++17 we can implement that!

Reply:
Oh right, I missed that the mode where used in some template arguments. :-(

Reply:
The nice side effect of procrastination is that you get C++17... :-)

Reply:
NOTE: this PR targets SYCL-1.2.1, which is based on C++11. This suggestion can be applied in `SYCL-2020 branch` though.